tests/db_functions/json/test_json_set.py

```diff
@@ -0,0 +1,403 @@
+        # that key instead of merging them.
+        UserPreferences.objects.update(
+            settings=JSONSet(
+                "settings",
+                theme={
+                    "type": "dark",
+                    "background": {
+                        "color": {"gradient-1": "black", "gradient-2": "grey"},
+                        "opacity": 0.5,
+                    },
+                },
+            )
+        )
+        user_preferences = UserPreferences.objects.get(pk=user_preferences.pk)
+        self.assertEqual(
+            user_preferences.settings,
+            {
+                "theme": {
+                    "type": "dark",
+                    "background": {
+                        "color": {"gradient-1": "black", "gradient-2": "grey"},
+                        "opacity": 0.5,
+                    },
+                },
+                "notifications": False,
+            },
+        )
+
+    def test_set_single_key_with_list(self):
+        user_preferences = UserPreferences.objects.create(
+            settings={"colors": ["black", "blue", "red"], "notifications": True}
+        )
+        UserPreferences.objects.update(
+            settings=JSONSet("settings", colors=["yellow", "green"])
+        )
+        user_preferences = UserPreferences.objects.get(pk=user_preferences.pk)
+        self.assertEqual(
+            user_preferences.settings,
+            {"colors": ["yellow", "green"], "notifications": True},
+        )
+
+    def test_set_single_key_with_list_using_index(self):
+        user_preferences = UserPreferences.objects.create(
+            settings={"colors": ["black", "blue", "red"], "notifications": True}
+        )
+        # Update using an existing index.
+        UserPreferences.objects.update(settings=JSONSet("settings", colors__1="yellow"))
+        user_preferences = UserPreferences.objects.get(pk=user_preferences.pk)
+        self.assertEqual(
+            user_preferences.settings,
+            {"colors": ["black", "yellow", "red"], "notifications": True},
+        )
+
+        # Update using an index that is equal to the array length. It should append the
+        # value to the array.
+        UserPreferences.objects.update(settings=JSONSet("settings", colors__3="green"))
+        user_preferences = UserPreferences.objects.get(pk=user_preferences.pk)
+        self.assertEqual(
+            user_preferences.settings,
+            {"colors": ["black", "yellow", "red", "green"], "notifications": True},
+        )
+
+        # Update using a nonexistent index that is greater than the array length.
+        UserPreferences.objects.update(settings=JSONSet("settings", colors__10="white"))
+        user_preferences = UserPreferences.objects.get(pk=user_preferences.pk)
+
+        # It should append the value to the array if the database supports it.
+        if not connection.features.json_set_array_append_requires_length_as_index:
+            if connection.vendor == "oracle":
```

Comment:
What happens here is that we're testing the behaviour of `JSON_SET` when updating an array with an index that's out-of-range. For example, with JSON `{"items": ["a","b","c"]}`.

If we do `JSONSet(field, items__3='d')`, this results in `{"items": ["a","b","c","d"]}` on all databases.

If we do `JSONSet(field, items__4='d')`, however, this gives different results:
- On SQLite, `{"items": ["a","b","c"]}`. The append does not happen.
- On Oracle, `{"items": ["a","b","c",null,"d"]}`. It fills out the "gap" with `null`, before appending `"d"` to the correct index.
- On other databases, `{"items": ["a","b","c","d"]}`. It just appends to the last index of the array.

We weren't sure if this is worth adding another feature flag, so we kind of combined Oracle and SQLite's behaviour into a single `json_set_array_append_requires_length_as_index` flag, with the base assumption being that the append does not happen, and Oracle being the outlier.

Perhaps, we should only do the full array assertion inside a vendor check for both `oracle` and `sqlite`. Then, add a general assertion that `field["items"][3]` does not equal `"d"` – this is true for both Oracle and SQLite, and we still leave room for external DB backends which may have the flag also set to true but we don't know the exact behaviour.